symbolic-il2cpp/src/usymlite.rs

```diff
@@ -0,0 +1,180 @@
+#[repr(C)]
+struct UsymLiteLine {
+    address: u64,
+    /// Reference to the managed source file name in the string table.
+    filename: u32,
+    /// Managed line number.
+    line: u32,
+}
+
+pub struct UsymLiteSymbols<'a> {
+    header: &'a UsymLiteHeader,
+    lines: &'a [UsymLiteLine],
+    /// The string table.
+    ///
+    /// This is a large slice of bytes with null-terminated C strings.
+    string_table: &'a [u8],
+}
+
+impl<'a> UsymLiteSymbols<'a> {
+    const MAGIC: &'static [u8] = b"sym-";
+
+    pub fn parse(buf: &'a [u8]) -> Result<UsymLiteSymbols<'a>> {
+        if buf.as_ptr().align_offset(8) != 0 {
+            // Alignment only really matters for performance really.
+            return Err(Error::msg("Data buffer not aligned to 8 bytes"));
+        }
+        if buf.len() < mem::size_of::<UsymLiteHeader>() {
+            return Err(Error::msg("Data smaller than UsymLiteHeader"));
+        }
+        if buf.get(..4) != Some(Self::MAGIC) {
+            return Err(Error::msg("Wrong magic number"));
+        }
+
+        // SAFETY: We checked the buffer is large enough above.
+        let header = unsafe { &*(buf.as_ptr() as *const UsymLiteHeader) };
+        if header.version != 2 {
+            return Err(Error::msg("Unknown version"));
+        }
+        let line_count: usize = header.line_count.try_into()?;
+        let stringtable_offset =
+            mem::size_of::<UsymLiteHeader>() + line_count * mem::size_of::<UsymLiteLine>();
+        if buf.len() < stringtable_offset {
+            return Err(Error::msg("Data smaller than number of line records"));
+        }
+
+        // SAFETY: We checked the buffer is at least the size_of::<UsymLiteHeader>() above.
+        let lines_ptr = unsafe { buf.as_ptr().add(mem::size_of::<UsymLiteHeader>()) };
+
+        // SAFETY: We checked the buffer has enough space for all the line records above.
+        let lines = unsafe {
+            let lines_ptr: *const UsymLiteLine = lines_ptr.cast();
+            let lines_ptr = ptr::slice_from_raw_parts(lines_ptr, line_count);
+            lines_ptr
+                .as_ref()
+                .ok_or(Error::msg("lines_ptr was null pointer!"))
+        }?;
+
+        let stringtable = buf
+            .get(stringtable_offset..)
+            .ok_or(Error::msg("No string table found"))?;
+        if stringtable.last() != Some(&0u8) {
+            return Err(Error::msg("String table does not end in NULL byte"));
+        }
+
+        Ok(Self {
+            header,
+            lines,
+            string_table: stringtable,
+        })
+    }
+
+    /// Returns a string from the string table at given offset.
+    ///
+    /// Offsets are as provided by some [`UsymLiteHeader`] and [`UsymLiteLine`] fields.
+    fn get_string(&self, offset: u32) -> Option<&'a CStr> {
+        // Panic if size_of::<usize>() < size_of::<u32>().
+        let offset: usize = offset.try_into().unwrap();
+        if offset >= self.string_table.len() {
```

Comment:
nit: i'm sure that inline type hints mitigate this problem, but `string_table` here feels like a real misnomer especially when used in this context because this comparison looks wrong at a glance: it looks as if you're checking the total number of entries in the string table against the offset, instead of the length of the table's raw contents.

Reply:
So I thought about this and ended up renaming the `string_table` variable in the **usym** module, but not here.

Reason is that a string table really is a very specific data structure, and it's a bunch of bytes with the occasional NULL byte.  So to me this is not a misnomer at all and the fact `string_table.len()` returns the size in bytes is entirely expected.

All of this is another way of saying that I understand what your reservation is here, but I entirely disagree 😄

Reply:
i won't block the PR on this, but i'm a bigger fan of the new name you used in `usym`. do you have a link to this definition of a string table (stringtable?) that you're referring to? i've got the feeling that this might be a bit of an esoteric definition, but i could be wrong. all i could find was windows documentation for a very specific structure used in windows apps development.

for me, i think the most common/prolific definition of a string table is closer to a mapping-type structure filled with strings, built off of the common usage of "table" as an alias or shortened form of "lookup table". given that, it makes this name really confusing as it seems to use a subtly different definition. this does kind of build a case against using "string table" to avoid confusion, but "stringtable" feels unnatural to me, even as a compound noun. it simply looks like a weird form of "string table" to me, but i don't really have the linguistic vocabulary to explain why that's the case.

Reply:
Err, fine.  Section 7.5.5 of the Dwarf 5 standard has a "string" entry which describes this (search for DW_FORM_strp).  LLVM mentions this in a header file: https://llvm.org/doxygen/StringTableBuilder_8h_source.html.

Reply:
Of course this is a domain-specific term in the debugging world.  But I believe the code here is part of this world and using domain-specific terms is fine.